byterun/caml/roots.h

```diff
@@ -35,10 +35,48 @@ CAMLextern void caml_do_local_roots (scanning_action, value *, value *,
 CAMLextern void caml_do_local_roots(scanning_action f, char * bottom_of_stack,
                                     uintnat last_retaddr, value * gc_regs,
                                     struct caml__roots_block * local_roots);
-#endif
+#endif /* NATIVE_CODE */
 
 CAMLextern void (*caml_scan_roots_hook) (scanning_action);
 
+
+/* Hooks to profile systhreads. */
+
+#define ST_NEXT_HOOK(thread_id,bottom,retaddr) \
+  if( st_next_hook != NULL ) st_next_hook(thread_id,bottom,retaddr)
+CAMLextern void (*st_next_hook)(value thread_id, char* bottom,
+                                        uintnat retaddr);
+/*
+ ST_HOOK_INIT: corresponding thread is the first thread of the
+    system. All preceeding allocations were done by it !
+ ST_HOOK_YIELD: corresponding thread is not going to allocate
+    anymore until scheduled again.
+ ST_HOOK_SCHEDULE: corresponding thread is being scheduled.
+    Further allocations are done by it !
+ ST_HOOK_STOP: corresponding thread is being terminated. It will
+    NEVER allocate again.
+ ST_HOOK_REINIT: corresponding thread is the first thread after a fork.
```

Comment:
I don't care what the numbers are, you must describe `INIT` and `REINIT` right next to each other.